### PR TITLE
ENH: Add ability to add hex encoded colors to outline items

### PR DESCRIPTION
**Pull request by mtd91429**
Add hexstring and basic named color parsing to `PdfWriter` method `add_outline_item()`

### CODE REVIEW

**Comment by MartinThoma** [comment]
I fear that this might overall do more harm than good:

1. Consistency: people can now expect those two types of color specifications in every place where PyPDF2 is accepting a color argument
2. Simplicity: it's clear what color a hex string maps to, but which shade of gray is "gray"?
3. Spelling: gray? Or Gray? GRAY? Grey?
4. Interface : I like to keep parameters simple - giving different types and different logic around those types to one parameter is something I try to avoid.


Those disadvantages can be avoided when you would create a Color helper class. That helper class could have class attributes that map to the hex value. Or even better they are returning a Color object which has a ".tohex()" function... This way you could also add methods like darken / mix / interpolate to combine colors.

Maybe there are even good 3rd party libraries that offer this?

**Comment by MartinThoma** [comment]
Or we really add just a minimal color class that has those color names as class attributes and the hey codes as values... That might be a good middle ground

**Comment by MartinThoma** [comment]
What do you think about https://pypi.org/project/colour/?

**Comment by mtd91429** [comment]
You raise some very good points, many of which I thought about but wasn't sure what was the best path forward. 

My overall aim goal was to reduce the clunkiness of the current color attribute within the outline item (with the potential to expand it to other aspects of the library). The RGB tuple from 0-1 is not very intuitive or common for encoding colors. At first, I simply wanted to add the hex values in basic 6 digit format (ex.,#800080). I think those are the most straight forward, common due to the web, and easier to work-with. 

But then I thought that sometimes, you don't want to look up a hex code for a basic color and thought using color names might be reasonable. For example, simply `color="blue"`. As I started to dig into named color formats, I realized that was a larger can of worms than I wanted. See the [Note in W3C docs](https://www.w3.org/TR/css-color-4/#named-colors), [xkcd color survery](https://xkcd.com/color/rgb/). For middle ground, I settled on the [basic 16 colors](https://en.wikipedia.org/wiki/Web_colors#Basic_colors) from the Web specification, then added Orange because it seemed like a notable exception for a basic color within the English language. 🤷‍♂️

> What do you think about https://pypi.org/project/colour/?

That was the logical step if we wanted to go "all in" on color features. But I don't think that colors are that central to the primary purpose of PyPDF2. I think it is overkill for the needs of the library. 

> a minimal color class that has those color names as class attributes and the hex codes as values

Yea, that may be a better idea. Keep it simple. 

I think in retrospect either 1) ditching the naming convention entirely and only supporting hexstrings or 2) use a minimal class to serve essentially as a dictionary for convenience.

**Comment by MartinThoma** [comment]
> At first, I simply wanted to add the hex values in basic 6 digit format (ex.,#800080). I think those are the most straight forward, common due to the web, and easier to work-with.

Yes!

> >What do you think about https://pypi.org/project/colour/?
>
> That was the logical step if we wanted to go "all in" on color features. But I don't think that colors are that central to the primary purpose of PyPDF2. I think it is overkill for the needs of the library.

I agree. I was thinking more about having a look at this library and pointing to it in the documentation :-) 

> 1) ditching the naming convention entirely and only supporting hexstrings or 2) use a minimal class to serve essentially as a dictionary for convenience

Both make sense to me. I think we could do both at the same time.

However, we would need to support both, the (r, g, b) tuple and the hex-string, for a while.

**Comment by mtd91429** [comment]
I slept on it and think that removing the named colors is probably the best route forward. It's more complicated than it is worth. Rather, I'm proposing to add the ability to parse hex strings instead. I've only implemented this for the outline items but think this could be expanded to other areas of the PDF where the color attribute is employed. For now, I think this is only within the outline item workflow, but is implied for other components (such as annotations) via a dictionary

File: PyPDF2/generic.py [hunk 1]
```diff
@@ -2020,30 +2020,29 @@ def create_string_object(
...
             {NameObject("/C"): ArrayObject([FloatObject(c) for c in color])}
         )
-
-    format_flag = 0
-    if italic:
-        format_flag += 1
-    if bold:
-        format_flag += 2
-    if format_flag:
+    if italic | bold:
```
Comment (MartinThoma) [inline PyPDF2/generic.py:2040]:
```suggestion
    if italic or bold:
```

File: tests/test_writer.py [hunk 2]
```diff
@@ -629,3 +631,25 @@ def test_deprecate_bookmark_decorator():
...
+    writer.add_outline_item("Second Outline Item", pagenum=3, color="#800080")
+    writer.add_outline_item("Third Outline Item", pagenum=4, color=purple_rgb)
+
+    target = "tmp-named-color-outline.pdf"
+    with open(target, "wb") as f:
+        writer.write(f)
+
+    reader2 = PdfReader(target)
+    for outline_item in reader2.outline:
+        # convert float to string b/c of mutability
```
Comment (MartinThoma) [inline tests/test_writer.py:651]:
```suggestion
        # convert float to string because of mutability
```

**Comment by MartinThoma** [comment]
@mtd91429 Thank you for all the work you continuously put into PyPDF2 :hugs:

**Pull request merged**